hibernate-core/src/main/java/org/hibernate/query/Query.java

```diff
@@ -37,63 +37,41 @@
 import jakarta.persistence.TypedQuery;
 
 /**
- * Represents an HQL/JPQL query or a compiled Criteria query.  Also acts as the Hibernate
- * extension to the JPA Query/TypedQuery contract
+ * Represents an HQL or Criteria query.
+ *
+ * Hibernate extension of {@link TypedQuery}, adapting to the
+ * {@link org.hibernate.query.CommonQueryContract} hierarchy
  *
  * @author Gavin King
  * @author Steve Ebersole
  *
  * @param <R> The query result type, for typed queries, or {@code Object} for untyped queries
  */
 @Incubating
-public interface Query<R> extends TypedQuery<R>, CommonQueryContract {
-	/**
-	 * Get the QueryProducer this Query originates from.  Generally speaking,
-	 * this is the Session/StatelessSession that was used to create the Query
-	 * instance.
-	 *
-	 * @return The producer of this query
-	 */
-	SharedSessionContract getSession();
-
-	/**
-	 * Get the query string.  Note that this may be {@code null} or some other
-	 * less-than-useful return because the source of the query might not be a
-	 * String (e.g., a Criteria query).
-	 *
-	 * @return the query string.
-	 */
-	String getQueryString();
-
-	/**
-	 * Apply the given graph using the given semantic
-	 *
-	 * @param graph The graph the apply.
-	 * @param semantic The semantic to use when applying the graph
-	 *
-	 * @return this - for method chaining
-	 */
-	Query<R> applyGraph(@SuppressWarnings("rawtypes") RootGraph graph, GraphSemantic semantic);
+public interface Query<R> extends UntypedQuery, JpaQuery, TypedQuery<R> {
```

Comment:
> I believe you added `UntypedQuery` because you wanted to stress that a query is untyped, but Java (for better or worse) already provides a way to do that: raw types.

@yrodiere it's a little more subtle than that. The issue is that a raw `Query` not only erases the (conceptually covariant) type parameter that occurs in the return types of the operations of `Query`, but, unfortunately, also all contravariant type arguments that occur in parameter types of the operations. 

This means, in our context, that you get all sorts of nasty and completely bogus warnings when you call operations of `Query`, including `setParameter()` and `addEntity()`&mdash;you're at risk of a warning anytime you call any operation with a generic type instantiation in its schema.

So it's not quite as simple as "oh, this is just the same as a raw type".

Reply:
Ok, but wouldn't this solve the problem just as well?

> I would suggest to make `SelectionQuery` parameterized (`SelectionQuery<R>`), add parameters whevever it makes sense (e.g. `Query<R> extends SelectionQuery<R>`), and use `SelectionQuery<?>` for legacy, untyped queries (e.g. `SelectionQuery<?> createdUntypedQuery(...)`. That would simplify your inheritance tree, and we really need all the simplifying we can get here.

I.e. making `SelectionQuery` parameterized and using `SelectionQuery<?>` (or `SelectionQuery<Object>` like you suggested elsewhere) in APIs. No raw types, no problems.

Reply:
Yeah, apologies: I was going to let Steve respond to that, but I read it and it didn't sound crazy or obviously unworkable to me. But without actually trying it out I can't say for sure if it works or not.

Reply:
> This looks weird: `Query<R> extends UntypedQuery, [...] TypedQuery<R>`.

Hm, not sure where you are seeing that.  I have:

```
public interface SelectionQuery extends CommonQueryContract {
    ...
}

public interface UntypedQuery extends SelectionQuery {
    ...
}
```

> 
> I believe you added `UntypedQuery` because you wanted to stress that a query is untyped, but Java (for better or worse) already provides a way to do that: raw types.
> 
> Plus, IMO a contract only makes sense if it offers operations. `UntypedQuery` is basically the same as `SelectionQuery`, but with a different name. Anywhere someone would use `UntypedQuery`, they could use `SelectionQuery` instead. So... I'm not sure it makes sense to ever use `UntypedQuery` in an application? Unless they really want to make sure their query is untyped, but why would they want that?

I'm ok with dropping `UntypedQuery` altogether in favor of `SelectionQuery`.

Initially `UntypedQuery` hooked in to the JPA Query hierarchy whereas `SelectionQuery` did not.  Forgot I changed that.  The intention being that users could treat `UntypedQuery` as `jakarta.persistence.Query`.  But you are right, once I dropped `UntypedQuery extends jakarta.persistence.Query` it became a moot point.

Ultimately I guess it comes down to how we want to handle `Query#createQuery(String)`


> I would suggest to make `SelectionQuery` parameterized (`SelectionQuery<R>`), add parameters whevever it makes sense (e.g. `Query<R> extends SelectionQuery<R>`), and use `SelectionQuery<?>` for legacy, untyped queries (e.g. `SelectionQuery<?> createdUntypedQuery(...)`. That would simplify your inheritance tree, and we really need all the simplifying we can get here.

But that's not an untyped (to me == raw) query really

Reply:
> I believe you added `UntypedQuery` because you wanted to stress that a query is untyped, but Java (for better or worse) already provides a way to do that: raw types.

No, I did it to somewhat mirror JPA's own Query hierarchy - just with (imo ofc) better naming

Reply:
> we really need all the simplifying we can get here.

Meh - i don't think its all that complicated really.  Yes the covariant stuff blows, but well... Java

Reply:
> > This looks weird: `Query<R> extends UntypedQuery, [...] TypedQuery<R>`.
> 
> Hm, not sure where you are seeing that. I have:

See above my comment. It's in the line I annotated.



> I'm ok with dropping `UntypedQuery` altogether in favor of `SelectionQuery`.

:+1: 

> But that's not an untyped (to me == raw) query really

Sure, it's not a raw type. I'd say it's still untyped since we don't promise that the returned hits will have a specific type, but YMMV.

In any case, I'd argue it's better than a raw query type:

* It accurately expresses the idea that "we don't know what type of hits this query returns" (just like a raw type would)
* It avoids the kind of problems we get from raw types (see [Gavin's comment above](https://github.com/hibernate/hibernate-orm/pull/4642#discussion_r785806689))
* It provides more useful compile-time checks to users when they incorrectly the query returns a specific type: for example `List<MyEntity> list = session.createUntypedQuery(...).getResultList();` will fail to compile, instead of just raising a warning. People will need to acknowledge that their query returns... untyped... entities and handle that.

Now, if you really want raw types for some reason, I guess you can use a raw `SelectionQuery`... That could work too, with the caveats mentioned by Gavin in his comment.

Reply:
AFAICS this was addressed by using non-raw return types in `SelectionQuery`, but `SelectionQuery` itself still isn't generic. Well, I guess time will tell if this was a good idea.